externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala

```diff
@@ -0,0 +1,366 @@
+        val allSchemas = listAllNamespaces(catalog, rootSchema)
+        allSchemas
+    }
+  }
+
+  private def listNamespacesWithPattern(
+      catalog: CatalogPlugin,
+      schemaPattern: String): Array[Array[String]] = {
+    listAllNamespaces(catalog).filter { ns =>
+      val quoted = ns.map(quoteIfNeeded).mkString(".")
+      schemaPattern.r.pattern.matcher(quoted).matches()
+    }.distinct
+  }
+
+  private def getSchemasWithPattern(catalog: CatalogPlugin, schemaPattern: String): Seq[String] = {
+    val p = schemaPattern.r.pattern
+    listAllNamespaces(catalog).flatMap { ns =>
+      val quoted = ns.map(quoteIfNeeded).mkString(".")
+      if (p.matcher(quoted).matches()) {
+        Some(quoted)
+      } else {
+        None
+      }
+    }.distinct
+  }
+
+  // ///////////////////////////////////////////////////////////////////////////////////////////////
+  //                                        Table & View                                         //
+  // ///////////////////////////////////////////////////////////////////////////////////////////////
+
+  def getCatalogTablesOrViews(
+      spark: SparkSession,
+      catalogName: String,
+      schemaPattern: String,
+      tablePattern: String,
+      tableTypes: Set[String],
+      ignoreTableProperties: Boolean = false): Seq[Row] = {
+    val catalog = getCatalog(spark, catalogName)
+    val namespaces = listNamespacesWithPattern(catalog, schemaPattern)
+    catalog match {
+      case builtin if builtin.name() == SESSION_CATALOG =>
+        val catalog = spark.sessionState.catalog
+        val databases = catalog.listDatabases(schemaPattern)
+
+        def isMatchedTableType(tableTypes: Set[String], tableType: String): Boolean = {
+          val typ = if (tableType.equalsIgnoreCase(VIEW)) VIEW else TABLE
+          tableTypes.exists(typ.equalsIgnoreCase)
+        }
+
+        databases.flatMap { db =>
+          val identifiers = catalog.listTables(db, tablePattern, includeLocalTempViews = false)
+          catalog.getTablesByName(identifiers)
+            .filter(t => isMatchedTableType(tableTypes, t.tableType.name)).map { t =>
+              val typ = if (t.tableType.name == VIEW) VIEW else TABLE
+              Row(
+                catalogName,
+                t.database,
+                t.identifier.table,
+                typ,
+                t.comment.getOrElse(""),
+                null,
+                null,
+                null,
+                null,
+                null)
+            }
+        }
+      case tc: TableCatalog =>
+        val tp = tablePattern.r.pattern
+        val identifiers = namespaces.flatMap { ns =>
+          tc.listTables(ns).filter(i => tp.matcher(quoteIfNeeded(i.name())).matches())
+        }
+        identifiers.map { ident =>
+          // TODO: restore view type for session catalog
+          val comment = if (ignoreTableProperties) ""
+          else tc.loadTable(ident).properties().getOrDefault(TableCatalog.PROP_COMMENT, "")
```

Comment:
It was reported by dozen users that metadata refresh takes a long time, let's add the comment to the code to make the learner easy to understand.

```suggestion
          val comment = if (ignoreTableProperties) ""
          else { // load table is a time consuming operation
            tc.loadTable(ident).properties().getOrDefault(TableCatalog.PROP_COMMENT, "")
          }
```

Reply:
Adopted.